modepy/test/test_matrices.py

```diff
@@ -118,6 +118,70 @@ def test_tensor_product_diag_mass_matrix(shape: mp.Shape) -> None:
                 assert np.max(err) < 1e-14
 
 
+@pytest.mark.parametrize("shape_cls", [mp.Hypercube, mp.Simplex])
+@pytest.mark.parametrize("dim", [1, 2, 3])
+@pytest.mark.parametrize("order", [0, 1, 2, 4])
+@pytest.mark.parametrize("nodes_on_bdry", [False, True])
+@pytest.mark.parametrize("test_weak_d_dr", [False, True])
+def test_bilinear_forms(
+            shape_cls: type[mp.Shape],
+            dim: int,
+            order: int,
+            nodes_on_bdry: bool,
+            test_weak_d_dr: bool
+        ) -> None:
+    shape = shape_cls(dim)
+    space = mp.space_for_shape(shape, order)
+
+    quad_space = mp.space_for_shape(shape, 2*order)
+    quad = mp.quadrature_for_space(quad_space, shape)
+
+    if isinstance(shape, mp.Hypercube) or shape == mp.Simplex(1):
+        if nodes_on_bdry:
+            nodes = mp.legendre_gauss_lobatto_tensor_product_nodes(
+                shape.dim, order,
+            )
+        else:
+            nodes = mp.legendre_gauss_tensor_product_nodes(shape.dim, order)
+    elif isinstance(shape, mp.Simplex):
+        if nodes_on_bdry:
+            nodes = mp.warp_and_blend_nodes(shape.dim, order)
+        else:
+            nodes = mp.VioreanuRokhlinSimplexQuadrature(order, shape.dim).nodes
+    else:
+        raise AssertionError()
```

Comment:
`mp.edge_clustered_nodes_for_space`?

Reply:
Updated. Though, for the case where `nodes_on_bdry` is false, we still have to do some checking unless I'm overlooking something.